### PR TITLE
FEATURE: Add dark mode optional logos for categories

### PR DESCRIPTION
**Pull request by cernikkk**
@pmusaraj suggested I open this draft because while working on this feature, I ran into a problem I need help with.

As far as I know, the feature "works", and the icons change depending on the theme as expected.

The problem I have is the `style="--aspect-ratio"` attribute applied to the parent div of the logo.

This attribute was originally defined inside `cdn-img`, but since I had to use that component inside a picture tag, I had to move it to `category-logo`. It is also applied through the following code inside `category-logo.js` (this is the part I don't fully understand):

```javascript
  @discourseComputed("width", "height")
  style(width, height) {
    if (width && height) {
      return htmlSafe(`--aspect-ratio: ${width / height};`);
    }
  },
```

Since `style="--aspect-ratio"` is only set when the component is called with "width" and "height", I quickly run into a loophole. As far as I understand, it can only be solved by modifying the behavior of the previously mentioned code, which I tried with no luck.

### CODE REVIEW

File: db/migrate/20140627193814_add_images_to_categories.rb
Comment (pmusaraj) [inline db/migrate/20140627193814_add_images_to_categories.rb:6]: This won't work. In Rails, once a migration has been run on a database, it won't be run again. Thus, we can't add actions to existing migrations. All the changes that you need to make to the database structure need to go in a new migration file. 

This applies to all the changes under `db/migrations`. I think that for this PR, we just need to add a `uploaded_logo_dark` column in a new migration. 

See https://edgeguides.rubyonrails.org/active_record_migrations.html for a more in-depth look at Rails migrations.

File: app/assets/javascripts/discourse/app/templates/components/categories-boxes-with-topics.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/templates/components/categories-boxes-with-topics.hbs:8]: `@alt=""` can be removed here, as far as I can tell, it doesn't do anything. (This wasn't great previously either, but since the `CategoryLogo` component is not an image at all, you don't need to pass that empty string.

File: app/assets/javascripts/discourse/app/templates/components/category-logo.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/templates/components/category-logo.hbs:1]: You probably don't need to check `this.category`, it should always be provided.

File: app/assets/javascripts/discourse/app/templates/components/category-logo.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/templates/components/category-logo.hbs:10]: I'm not sure we need this last `else` case. It's odd to upload _just_ a dark logo.

File: app/assets/javascripts/discourse/app/templates/navigation/category.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/templates/navigation/category.hbs:4]:
```suggestion
  {{#if this.category.uploaded_logo.url}}
```

I think we should keep a simpler rule here (and elsewhere): 

- only category logo is required for the display of the `CategoryLogo` component
- the dark logo should be only displayed as an alternative, that is, if the regular `category.uploaded_logo` is empty, then dark logo is not included in the HTML

File: app/views/list/list.erb
Comment (pmusaraj) [inline app/views/list/list.erb:146]: I don't think we need to make this change, the crawlable meta data should use the regular logo.

File: lib/oneboxer.rb
Comment (pmusaraj) [inline lib/oneboxer.rb:385]: I don't think we need this at this point.

**Review by pmusaraj — CHANGES_REQUESTED** [review]
The main requested change here is the one about migrations. Otherwise, this looks like a great start.

File: db/migrate/20140627193814_add_images_to_categories.rb
Comment (cernikkk) [reply db/migrate/20140627193814_add_images_to_categories.rb]: Well, seeing it now, what I did there was atrocious!
I reverted the changes and created a new migration file now.
I also had to add `uploaded_logo_dark_id` in it.

File: app/assets/javascripts/discourse/app/components/category-logo.js
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/components/category-logo.js:3]: Can you try a template only Glimmer component here?

Basically it would be to replace the contents of this file with: 

```
import templateOnly from "@ember/component/template-only";
export default templateOnly();
```

Glimmer is the new way Ember components are done.

File: app/assets/javascripts/discourse/app/templates/components/categories-boxes-with-topics.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/templates/components/categories-boxes-with-topics.hbs:8]: Minor nitpick, but it doesn't look like the "logo" class is being used, we should remove it. (It's too generic anyway, might clash with other, unrelated styles.)

File: app/assets/javascripts/discourse/app/templates/navigation/category.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/templates/navigation/category.hbs:5]: I don't see the "category-logo" class being used either, that also can be omitted.

File: config/locales/client.en.yml
Comment (pmusaraj) [inline config/locales/client.en.yml:3452]:
```suggestion
      logo_dark: "Dark Mode Category Logo Image"
```

File: db/migrate/20161025083648_fix_category_logo_and_background_urls.rb
Comment (pmusaraj) [inline db/migrate/20161025083648_fix_category_logo_and_background_urls.rb:10]: We don't need this change on this old migration.

File: db/migrate/20221004122343_add_dark_mode_logo_to_categories.rb
Comment (pmusaraj) [inline db/migrate/20221004122343_add_dark_mode_logo_to_categories.rb:5]: Is this `logo_dark_url` column being used?

File: app/assets/stylesheets/embed.scss
Comment (pmusaraj) [inline app/assets/stylesheets/embed.scss:156]: The embed stylesheet is a completely other part of the app, this change shouldn't affect it at all. So we shouldn't modify this stylesheet.

File: app/assets/stylesheets/common/base/header.scss
Comment (pmusaraj) [inline app/assets/stylesheets/common/base/header.scss:29]: This shouldn't be affected by this PR either (category logos have nothing to do with the header).

File: app/assets/javascripts/discourse/app/templates/components/category-logo.hbs
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/templates/components/category-logo.hbs:1]: Maybe we should change this to 
```suggestion
<div class="category-logo aspect-image">
```

What do you think? It would make for more reasonable style names in the stylesheets.

**Review by pmusaraj — COMMENTED** [review]
Just a few more comments, getting really close here.

File: app/assets/javascripts/discourse/app/templates/components/category-logo.hbs
Comment (cernikkk) [reply app/assets/javascripts/discourse/app/templates/components/category-logo.hbs:1]: I don't quite get the "Suggested change". Do you mean adding `category-logo` back with a better name or renaming `aspect-image` something more useful? `category-logo-container` maybe? I don't think we can get much more specific than that.

File: db/migrate/20221004122343_add_dark_mode_logo_to_categories.rb
Comment (pmusaraj) [reply db/migrate/20221004122343_add_dark_mode_logo_to_categories.rb:5]: Maybe you forgot to push that commit? That line is still there. It also needs to be removed from the schema information in the `category.rb` model file.

Comment (cernikkk) [reply db/migrate/20221004122343_add_dark_mode_logo_to_categories.rb]: I was waiting for your response on the other topic to do so. My comment was a little misleading, though. I'll try to be more clear in the future.
All changes are pushed now.

**Review by pmusaraj — APPROVED** [review]
Cool, thanks for all the fixes. I will merge this shortly (once tests pass). 

Thanks @cernikkk!

**Comment by cernikkk** [comment]
Thanks you too for the help!

**Pull request merged**
